File: sdk/csharp/libraries/microsoft.bot.solutions/Authentication/MultiProviderAuthDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Connector.Authentication;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Solutions.Responses;

namespace Microsoft.Bot.Solutions.Authentication
{
    /// <summary>
    /// Provides the ability to prompt for which Authentication provider the user wishes to use.
    /// </summary>
    public class MultiProviderAuthDialog : ComponentDialog
    {
        private static readonly string[] AcceptedLocales = new string[] { "en", "de", "es", "fr", "it", "zh" };
        private string _selectedAuthType = string.Empty;
        private List<OAuthConnection> _authenticationConnections;
        private ResponseManager _responseManager;
        private AppCredentials _oauthCredentials;

        public MultiProviderAuthDialog(
            List<OAuthConnection> authenticationConnections,
            List<OAuthPromptSettings> promptSettings = null,
            AppCredentials oauthCredentials = null)
            : base(nameof(MultiProviderAuthDialog))
        {
            _authenticationConnections = authenticationConnections ?? throw new ArgumentNullException(nameof(authenticationConnections));
            _oauthCredentials = oauthCredentials;
            _responseManager = new ResponseManager(
                AcceptedLocales,
                new AuthenticationResponses());

            var firstStep = new WaterfallStep[]
            {
                FirstStepAsync,
            };

            var authSteps = new WaterfallStep[]
            {
                PromptForProviderAsync,
                PromptForAuthAsync,
                HandleTokenResponseAsync,
            };

            AddDialog(new WaterfallDialog(DialogIds.FirstStepPrompt, firstStep));

            if (_authenticationConnections != null &&
                _authenticationConnections.Count > 0 &&
                _authenticationConnections.Any(c => !string.IsNullOrWhiteSpace(c.Name)))
            {
                for (int i = 0; i < _authenticationConnections.Count; ++i)
                {
                    var connection = _authenticationConnections[i];

                    foreach (var locale in AcceptedLocales)
                    {
                        // We ignore placeholder connections in config that don't have a Name
                        if (!string.IsNullOrWhiteSpace(connection.Name))
                        {
                            AddDialog(GetLocalizedDialog(locale, connection.Name, promptSettings?[i]));
                        }
                    }
                }

                AddDialog(new WaterfallDialog(DialogIds.AuthPrompt, authSteps));
                AddDialog(new ChoicePrompt(DialogIds.ProviderPrompt));
            }
            else
            {
                throw new ArgumentNullException(nameof(authenticationConnections));
            }
        }

        // Validators
        protected Task<bool> TokenResponseValidatorAsync(PromptValidatorContext<Activity> pc, CancellationToken cancellationToken)
        {
            var activity = pc.Recognized.Value;
            if (activity != null &&
               ((activity.Type == ActivityTypes.Event && activity.Name == TokenEvents.TokenResponseEventName) ||
               (activity.Type == ActivityTypes.Invoke && activity.Name == "signin/verifyState")))
            {
                return Task.FromResult(true);
            }

            return Task.FromResult(false);
        }

        private async Task<DialogTurnResult> FirstStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            return await stepContext.BeginDialogAsync(DialogIds.AuthPrompt).ConfigureAwait(false);
        }

        private async Task<DialogTurnResult> PromptForProviderAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            if (_authenticationConnections.Count == 1)
            {
                var result = _authenticationConnections.First().Name + "_" + CultureInfo.CurrentUICulture.TwoLetterISOLanguageName;
                return await stepContext.NextAsync(result).ConfigureAwait(false);
            }

            if (stepContext.Context.Adapter is IExtendedUserTokenProvider adapter)
            {
                var tokenStatusCollection = await adapter.GetTokenStatusAsync(stepContext.Context, _oauthCredentials, stepContext.Context.Activity.From.Id, null, cancellationToken).ConfigureAwait(false);

                var matchingProviders = tokenStatusCollection.Where(p => (bool)p.HasToken && _authenticationConnections.Any(t => t.Name == p.ConnectionName)).ToList();

                if (matchingProviders.Count == 1)
                {
                    var authType = matchingProviders[0].ConnectionName;
                    return await stepContext.NextAsync(authType, cancellationToken).ConfigureAwait(false);
                }

                if (matchingProviders.Count > 1)
                {
                    var choices = new List<Choice>();

                    foreach (var connection in matchingProviders)
                    {
                        choices.Add(new Choice
                        {
                            Action = new CardAction(ActionTypes.ImBack, connection.ConnectionName, value: connection.ConnectionName),
                            Value = connection.ConnectionName,
                        });
                    }

                    return await stepContext.PromptAsync(
                        DialogIds.ProviderPrompt,
                        new PromptOptions
                        {
                            Prompt = _responseManager.GetResponse(AuthenticationResponses.ConfiguredAuthProvidersPrompt),
                            Choices = choices,
                        },
                        cancellationToken).ConfigureAwait(false);
                }
                else
                {
                    var choices = new List<Choice>();

                    foreach (var connection in _authenticationConnections)
                    {
                        choices.Add(new Choice
                        {
                            Action = new CardAction(ActionTypes.ImBack, connection.Name, value: connection.Name),
                            Value = connection.Name,
                        });
                    }

                    return await stepContext.PromptAsync(
                        DialogIds.ProviderPrompt,
                        new PromptOptions
                        {
                            Prompt = _responseManager.GetResponse(AuthenticationResponses.AuthProvidersPrompt),
                            Choices = choices,
                        },
                        cancellationToken).ConfigureAwait(false);
                }
            }

            throw new Exception("The adapter doesn't support Token Handling.");
        }

        private async Task<DialogTurnResult> PromptForAuthAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            if (stepContext.Result is string selectedAuthType)
            {
                _selectedAuthType = selectedAuthType;
            }
            else if (stepContext.Result is FoundChoice choice)
            {
                _selectedAuthType = choice.Value;
            }

            return await stepContext.PromptAsync(_selectedAuthType, new PromptOptions(), cancellationToken).ConfigureAwait(false);
        }

        private async Task<DialogTurnResult> HandleTokenResponseAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            if (stepContext.Result is TokenResponse tokenResponse && !string.IsNullOrWhiteSpace(tokenResponse.Token))
            {
                var result = await CreateProviderTokenResponseAsync(stepContext.Context, tokenResponse).ConfigureAwait(false);

                return await stepContext.EndDialogAsync(result, cancellationToken).ConfigureAwait(false);
            }

            TelemetryClient.TrackEvent("TokenRetrievalFailure");
            return new DialogTurnResult(DialogTurnStatus.Cancelled);
        }

        private async Task<ProviderTokenResponse> CreateProviderTokenResponseAsync(ITurnContext context, TokenResponse tokenResponse)
        {
            var tokens = await GetTokenStatusAsync(context, context.Activity.From.Id).ConfigureAwait(false);
            var match = Array.Find(tokens, t => t.ConnectionName == tokenResponse.ConnectionName);

            return new ProviderTokenResponse
            {
                AuthenticationProvider = match.ServiceProviderDisplayName.GetAuthenticationProvider(),
                TokenResponse = tokenResponse,
            };
        }

        private async Task<TokenStatus[]> GetTokenStatusAsync(ITurnContext context, string userId, string includeFilter = null, CancellationToken cancellationToken = default(CancellationToken))
        {
            BotAssert.ContextNotNull(context);
            if (string.IsNullOrWhiteSpace(userId))
            {
                throw new ArgumentNullException(nameof(userId));
            }

            var tokenProvider = context.Adapter as IExtendedUserTokenProvider;
            if (tokenProvider != null)
            {
                return await tokenProvider.GetTokenStatusAsync(context, _oauthCredentials, userId, includeFilter, cancellationToken).ConfigureAwait(false);
            }
            else
            {
                throw new Exception("Adapter does not support IExtendedUserTokenProvider");
            }
        }

        private Task<bool> AuthPromptValidatorAsync(PromptValidatorContext<TokenResponse> promptContext, CancellationToken cancellationToken)
        {
            var token = promptContext.Recognized.Value;
            if (token != null && !string.IsNullOrWhiteSpace(token.Token))
            {
                return Task.FromResult(true);
            }

            var eventActivity = promptContext.Context.Activity.AsEventActivity();
            if (eventActivity != null && eventActivity.Name == TokenEvents.TokenResponseEventName)
            {
                promptContext.Recognized.Value = eventActivity.Value as TokenResponse;
                return Task.FromResult(true);
            }

            TelemetryClient.TrackEvent("AuthPromptValidatorAsyncFailure");
            return Task.FromResult(false);
        }

        private OAuthPrompt GetLocalizedDialog(string locale, string connectionName, OAuthPromptSettings settings)
        {
            var loginButtonActivity = _responseManager.GetResponse(AuthenticationResponses.LoginButton, locale);
            var loginPromptActivity = _responseManager.GetResponse(AuthenticationResponses.LoginPrompt, locale, new StringDictionary() { { "authType", connectionName } });
            settings = settings ?? new OAuthPromptSettings
            {
                ConnectionName = connectionName,
                Title = loginButtonActivity.Text,
                Text = loginPromptActivity.Text,
            };
            settings.OAuthAppCredentials = _oauthCredentials;

            return new OAuthPrompt(
                connectionName + "_" + locale,
                settings,
                AuthPromptValidatorAsync);
        }

        private static class DialogIds
        {
            public const string ProviderPrompt = "ProviderPrompt";
            public const string FirstStepPrompt = "FirstStep";
            public const string AuthPrompt = "AuthPrompt";
        }
    }
}

File: sdk/typescript/libraries/bot-solutions/src/authentication/multiProviderAuthDialog.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { BotFrameworkAdapter, TurnContext } from 'botbuilder';
import { Choice, ChoicePrompt, ComponentDialog, DialogTurnResult, DialogTurnStatus, FoundChoice,
    OAuthPrompt, PromptValidatorContext, WaterfallDialog, WaterfallStep, WaterfallStepContext,
    OAuthPromptSettings,
    Dialog } from 'botbuilder-dialogs';
import { TokenStatus } from 'botframework-connector';
import { ActionTypes, Activity, ActivityTypes, TokenResponse } from 'botframework-schema';
import { IOAuthConnection } from '../authentication';
import { ResponseManager } from '../responses';
import { TokenEvents } from '../tokenEvents';
import { AuthenticationResponses } from './authenticationResponses';
import { OAuthProviderExtensions } from './oAuthProviderExtensions';
import { IProviderTokenResponse } from './providerTokenResponse';
import { AppCredentials } from 'botframework-connector';

enum DialogIds {
    providerPrompt = 'ProviderPrompt',
    firstStepPrompt = 'FirstStep',
    authPrompt = 'AuthPrompt',
}

/**
 * Provides the ability to prompt for which Authentication provider the user wishes to use.
 */
export class MultiProviderAuthDialog extends ComponentDialog {
    private static readonly acceptedLocales: string[] = ['en', 'de', 'es', 'fr', 'it', 'zh'];
    private selectedAuthType: string = '';
    private authenticationConnections: IOAuthConnection[];
    private responseManager: ResponseManager;
    private oauthCredentials: AppCredentials | undefined;

    public constructor(
        authenticationConnections: IOAuthConnection[],
        locale: string,
        promptSettings: OAuthPromptSettings[],
        oauthCredentials?: AppCredentials
    ) {
        super(MultiProviderAuthDialog.name);

        if (authenticationConnections === undefined) { throw new Error('The value of authenticationConnections cannot be undefined'); }
        this.authenticationConnections = authenticationConnections;
        this.oauthCredentials = oauthCredentials;
        this.responseManager = new ResponseManager(
            MultiProviderAuthDialog.acceptedLocales,
            [AuthenticationResponses]
        );

        const firstStep: WaterfallStep[] = [
            this.firstStep.bind(this)
        ];

        const authSteps: WaterfallStep[] = [
            this.promptForProvider.bind(this),
            this.promptForAuth.bind(this),
            this.handleTokenResponse.bind(this)
        ];

        this.addDialog(new WaterfallDialog(DialogIds.firstStepPrompt, firstStep));

        if (this.authenticationConnections !== undefined && 
            this.authenticationConnections.length > 0 && 
            this.authenticationConnections.some((c: IOAuthConnection): boolean => c.name !== undefined && c.name.trim().length > 0)) {
                
            for (var i = 0; i < this.authenticationConnections.length; ++i) {
                let connection = this.authenticationConnections[i];

                MultiProviderAuthDialog.acceptedLocales.forEach((locale): void => {
                    this.addDialog(this.getLocalizedDialog(locale, connection.name, promptSettings[i]));
                });
            };

            this.addDialog(new WaterfallDialog(DialogIds.authPrompt, authSteps));
            this.addDialog(new ChoicePrompt(DialogIds.providerPrompt));
        } else {
            throw new Error('There is no authenticationConnections value');
        }
    }

    private getLocalizedDialog(locale: string, connectionName: string, settings: OAuthPromptSettings): Dialog {
        const loginButtonActivity: string = this.responseManager.getLocalizedResponse(AuthenticationResponses.loginButton, locale).text;
        const loginPromptActivity: string = this.responseManager.getLocalizedResponse(AuthenticationResponses.loginPrompt, locale, new Map<string, string>([['authType', connectionName]])).text;
        
        settings = settings || {
            connectionName: connectionName,
            title: loginButtonActivity,
            text: loginPromptActivity
        };

        return new OAuthPrompt(
            connectionName + '_' + locale,
            settings,
            this.authPromptValidator.bind(this)
        );
    }

    // Validators
    protected async tokenResponseValidator(pc: PromptValidatorContext<Activity>): Promise<boolean> {
        const activity: Activity | undefined = pc.recognized.value;
        if (activity !== undefined && 
            ((activity.type === ActivityTypes.Event && activity.name === TokenEvents.tokenResponseEventName) || 
            (activity.type === ActivityTypes.Invoke && activity.name === 'signin/verifyState'))) {
            return Promise.resolve(true);
        }

        return Promise.resolve(false);
    }

    private async firstStep(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
        return await stepContext.beginDialog(DialogIds.authPrompt);
    }

    private async promptForProvider(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
        if (this.authenticationConnections.length === 1) {
            const result: string = this.authenticationConnections[0].name + '_' + stepContext.context.activity.locale.split('-')[0];

            return await stepContext.next(result);
        }

        //PENDING: adapter could not be parsed to IExtendedUserTokenProvider as C# does
        const adapter: BotFrameworkAdapter = stepContext.context.adapter as BotFrameworkAdapter;
        if (adapter !== undefined) {
            const tokenStatusCollection: TokenStatus[] = await adapter.getTokenStatus(
                stepContext.context,
                stepContext.context.activity.from.id,
                undefined,
                this.oauthCredentials);

            const matchingProviders: TokenStatus[] = tokenStatusCollection.filter((p: TokenStatus): boolean => {
                return (p.hasToken || false) && this.authenticationConnections.some((t: IOAuthConnection): boolean => {
                    return t.name === p.connectionName;
                });
            });

            if (matchingProviders.length === 1) {
                const authType: string|undefined = matchingProviders[0].connectionName;

                return stepContext.next(authType);
            }

            if (matchingProviders.length > 1) {
                const choices: Choice[] = matchingProviders.map((connection: TokenStatus): Choice => {
                    const value: string = connection.connectionName || '';

                    return {
                        action: {
                            type: ActionTypes.ImBack,
                            title: value,
                            value: value
                        },
                        value: value
                    };
                });

                return stepContext.prompt(DialogIds.providerPrompt, {
                    prompt: this.responseManager.getResponse(AuthenticationResponses.configuredAuthProvidersPrompt, stepContext.context.activity.locale as string),
                    choices: choices
                });
            } else {
                const choices: Choice[] = this.authenticationConnections.map((connection: IOAuthConnection): Choice => {
                    return {
                        action: {
                            type: ActionTypes.ImBack,
                            title: connection.name,
                            value: connection.name
                        },
                        value: connection.name
                    };
                });

                return stepContext.prompt(DialogIds.providerPrompt, {
                    prompt: this.responseManager.getResponse(AuthenticationResponses.authProvidersPrompt, stepContext.context.activity.locale as string),
                    choices: choices
                });
            }
        }

        throw new Error('The adapter doesn\'t support Token Handling.');
    }

    private async promptForAuth(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
        if (typeof stepContext.result === 'string') {
            this.selectedAuthType = stepContext.result;
        } else {
            const choice: FoundChoice = stepContext.result as FoundChoice;
            if (choice !== undefined) {
                this.selectedAuthType = choice.value;
            }
        }

        return await stepContext.prompt(this.selectedAuthType, {});
    }

    private async handleTokenResponse(stepContext: WaterfallStepContext): Promise<DialogTurnResult> {
        const tokenResponse: TokenResponse = stepContext.result as TokenResponse;

        if (tokenResponse !== undefined && tokenResponse.token) {
            const result: IProviderTokenResponse = await this.createProviderTokenResponse(stepContext.context, tokenResponse);

            return await stepContext.endDialog(result);
        }

        this.telemetryClient.trackEvent({
            name: 'TokenRetrievalFailure'
        });

        return { status: DialogTurnStatus.cancelled };
    }

    private async createProviderTokenResponse(context: TurnContext, tokenResponse: TokenResponse): Promise<IProviderTokenResponse> {
        const tokens: TokenStatus[] = await this.getTokenStatus(context, context.activity.from.id);
        const match: TokenStatus|undefined = tokens.find((t: TokenStatus): boolean => t.connectionName === tokenResponse.connectionName);

        if (!match) {
            throw new Error('Token not found');
        }

        const response: IProviderTokenResponse = {
            authenticationProvider: OAuthProviderExtensions.getAuthenticationProvider(match.serviceProviderDisplayName || ''),
            tokenResponse: tokenResponse
        };

        return Promise.resolve(response);
    }

    private async getTokenStatus(context: TurnContext, userId: string, includeFilter?: string): Promise<TokenStatus[]> {
        if (context === undefined) {
            throw new Error('"context" undefined');
        }

        if (userId === undefined || userId.trim().length === 0) {
            throw new Error('"userId" undefined');
        }

        //PENDING: adapter could not be parsed to IExtendedUserTokenProvider as C# does
        const tokenProvider: BotFrameworkAdapter = context.adapter as BotFrameworkAdapter;
        if (tokenProvider !== undefined) {
            return await tokenProvider.getTokenStatus(context, userId, includeFilter, this.oauthCredentials);
        } else {
            throw new Error('Adapter does not support IExtendedUserTokenProvider');
        }
    }

    private async authPromptValidator(promptContext: PromptValidatorContext<TokenResponse>): Promise<boolean> {
        const token: TokenResponse|undefined = promptContext.recognized.value;
        if (token !== undefined && token.token !== undefined && token.token.trim().length > 0) {
            return Promise.resolve(true);
        }

        const eventActivity: Activity = promptContext.context.activity;
        if (eventActivity !== undefined && eventActivity.name === TokenEvents.tokenResponseEventName) {
            promptContext.recognized.value = eventActivity.value as TokenResponse;

            return Promise.resolve(true);
        }

        this.telemetryClient.trackEvent({
            name: 'AuthPromptValidatorAsyncFailure'
        });

        return Promise.resolve(false);
    }
}
